perceval/components/detector.py

```diff
@@ -26,9 +26,160 @@
 
-class Detector():
     def __init__(self):
+        super().__init__(1)
+
+    @property
+    @abstractmethod
+    def type(self) -> DetectionType:
+        """
+        Returns the detector type
+        """
+
+    @abstractmethod
+    def detect(self, theoretical_photons: int) -> BSDistribution or BasicState:
+        """
+        Returns a one mode Fock state distribution out of a theoretical photon count coming in the detector
+        """
+
+
+class BSLayeredPPNR(IDetector):
+    """
+    BSLayeredPPNR implements Pseudo Photon Number Resolving detection using layers of beam splitter plugged on
+    2**(number of layers) threshold detectors.
+
+    :param bs_layers: Number of beam splitter layers. Adding more layers enabled to detect
+    """
+
+    def __init__(self, bs_layers: int, reflectivity: float = 0.5):
+        assert isinstance(bs_layers, int) and bs_layers > 0,\
+            "Beam-splitter layers have to be a stricly positive integer"
+        assert 0 <= reflectivity <= 1, f"Reflectivity must be between 0 and 1 (got {reflectivity})"
+        super().__init__()
+        self.name = f"BS-PPNR{bs_layers}"
+        self._layers = bs_layers
+        self._r = reflectivity
+        self._cache = {}  # This cache records simulations for a given photon count to speed up computations
+
+    @property
+    def type(self) -> DetectionType:
+        return DetectionType.PPNR
+
+    def clear_cache(self):
+        self._cache = {}
+
+    def create_circuit(self) -> Circuit:
         """
-            definition of a detector
+        Creates the circuit to simulate PPNR with threshold detectors
         """
+        ppnr_circuit = Circuit(2 ** self._layers)
+        for l in range(self._layers):
+            perm_vector = list(range(0, 2**(l+1)-1, 2)) + list(range(1, 2**(l+1)-1, 2))
+            if len(perm_vector) > 1:
+                ppnr_circuit.add(0, PERM(perm_vector))
+            for m in range(0, 2**(l+1), 2):
+                ppnr_circuit.add(m, BS(BS.r_to_theta(self._r)))
+        return ppnr_circuit
+
+    def detect(self, theoretical_photons: int) -> BSDistribution or BasicState:
+        if theoretical_photons < 2:
+            return BasicState([theoretical_photons])
+
+        if theoretical_photons in self._cache:
+            return self._cache[theoretical_photons]
+
+        from perceval.backends import SLOSBackend
+        ppnr_circuit = self.create_circuit()
+        slos = SLOSBackend()
+        slos.set_circuit(ppnr_circuit)
+        slos.set_input_state(BasicState([theoretical_photons] + [0]*(ppnr_circuit.m - 1)))
+        dist = slos.prob_distribution()
+
+        output = BSDistribution()
+        for state, prob in dist.items():
+            state = state.threshold_detection()
+            output[BasicState([state.n])] += prob
+        self._cache[theoretical_photons] = output
+        return output
+
+
+class Detector(IDetector):
+    def __init__(self, p_multiphoton_detection: float = 1):
```

Comment:
Here, there is a risk that users use the __init__ method directly instead of the threshold(), pnr() and ppnr() functions. 
Maybe this class should still be abstract and be subclassed by the three detectors ?